### ISSUE TITLE
Electron 4 upgrade causes thin font rendering on MacOS 10.14+

### ISSUE THREAD

**Issue by connorlr** [atom/atom#20052]
The upgrade from Electron 3 to 4 significantly thins out the font rendering in the editor, making it much harder to read. This is a known issue with electron 4.

https://github.com/electron/electron/issues/19970

**Comment by Arcanemagus** [comment]
This was considered for the Electron 3 upgrade, but was reported to have reverted to the old behavior for the Electron 4 upgrade in https://github.com/atom/atom/issues/19356.

How does the rendering change if you toggle the _system_ setting "Use font smoothing when available"?

What about if you apply the following CSS in [your stylesheet](https://flight-manual.atom.io/using-atom/sections/basic-customization/#style-tweaks)?
```less
/* This will change how font-smoothing is applied to the entire Atom application */
atom-workspace {
  -webkit-font-smoothing: antialiased;
}

/* If you just want to change how text is rendered within the editor, use this instead: */
atom-text-editor {
  -webkit-font-smoothing: antialiased;
}
```

**Comment by brendanfalkowski** [comment]
Came here to open the same bug. 

Here's the appearance on a Dell 27" display (1x pixel density, not a 4k display) and OneDark theme.

All the straight edges in characters are fuzzed after updating to Atom 1.41.0.

— — — — — — — — — — 

## My old fix

Previously, I had to hack the stylesheet due this happening prior, and not finding a better fix:
https://discuss.atom.io/t/macos-mojave-font-rendering-change/59058

```
atom-text-editor {
    font-weight: 600;
}
```

![Screen Shot 2019-10-23 at 3 24 57 PM](https://user-images.githubusercontent.com/214924/67438796-4eb0d280-f5a9-11e9-96f1-13bbbbb60a65.png)

— — — — — — — — — — 

## My old fix + suggested new fix

With the fix suggested above added — still fuzzy:

```
atom-text-editor {
    font-weight: 600;
    -webkit-font-smoothing: antialiased;
}
```

![Screen Shot 2019-10-23 at 3 23 30 PM](https://user-images.githubusercontent.com/214924/67438727-1e693400-f5a9-11e9-8674-108e4c6cbb1b.png)

— — — — — — — — — — 

## Suggested new fix

Still fuzzy.

```
atom-text-editor {
    // font-weight: 600;
    -webkit-font-smoothing: antialiased;
}
```

![Screen Shot 2019-10-23 at 3 30 33 PM](https://user-images.githubusercontent.com/214924/67439053-12ca3d00-f5aa-11e9-94c2-59ea320ee16e.png)

— — — — — — — — — — 

## No fix, current Atom default

And with neither adjustment in place, this would be pretty miserable:

```
atom-text-editor {
    // font-weight: 600;
    // -webkit-font-smoothing: antialiased;
}
```

![Screen Shot 2019-10-23 at 3 27 27 PM](https://user-images.githubusercontent.com/214924/67438914-a4857a80-f5a9-11e9-821f-04184de339c5.png)


How does this pass QA? Everybody on HiDPI laptops?

— — — — — — — — — — 

My Mac has `use font smoothing when available` turned on (the default I think).

**Comment by leotada** [comment]
https://electronjs.org/releases/stable#end-of-support-for-4xy

**Comment by brendanfalkowski** [comment]
YOLO — new custom stylesheet

```
atom-workspace {
    // This doesn't work, IDK
    font-weight: 700;
}

atom-text-editor {
    font-weight: 700;
}
```

**Comment by victorhqc** [comment]
Is this related maybe?
https://github.com/atom/atom/issues/19356

**Comment by connorlr** [comment]
@Arcanemagus 

Default appearance (system font smoothing enabled)
![image](https://user-images.githubusercontent.com/50108264/67489385-7eaab500-f63f-11e9-8c5b-85abf68c5466.png)

Font smoothing disabled
![image](https://user-images.githubusercontent.com/50108264/67489546-c16c8d00-f63f-11e9-93f4-55e0b39a0a2f.png)

With `-webkit-font-smoothing: antialiased;`
![image](https://user-images.githubusercontent.com/50108264/67489633-e2cd7900-f63f-11e9-9fe1-8a95ff1c631b.png)

None of these options appear to have any effect on atom, though disabling font smoothing had a clear negative effect on fonts in other applications.

**Comment by brendanfalkowski** [comment]
@connorlr Guessing from your screenshot that's a HiDPI monitor. I'm pretty sure both macOS and Webkit forcefully disable antialiasing (regardless of settings or CSS you have) because the antialiasing effect is mute when the pixels are so dense. They can just render aliased and there's so much canvas to paint the shape it looks smooth to the eye regardless.

This should only affect a 1x monitor like a Dell U2717D (2560x1440). It won't affect any current Mac laptop/desktop monitor or external 4K monitor with double that resolution.

**Comment by connorlr** [comment]
@brendanfalkowski That's correct, I am on an LG Ultra HD monitor. I'm not sure what you mean by "This should only affect a 1x monitor" though. If you're talking about the webkit antialiasing option then I agree, but the overall issue of overly-thin font still applies regardless of resolution.

**Comment by brendanfalkowski** [comment]
A 2x or higher device *should* never actually have antialiasing enabled for text rendering. That's enforced by Apple and Webkit/Chrome because the extra computation required is wasted if the screen has such high pixel density. That's why it has no effect when you turn it on/off (it was never on).

I don't know a thing about Electron, so maybe this doesn't apply equally. It's possible they're thinning with another technique proprietary to Electron apps or another of the web techniques for thinning like text-shadow.

This type of fuzzing can also occur in Webkit when an element is promoted to a GPU-rendered layer (by using a 3D transform or setting will-change). If that box has an odd width, the GPU renders the whole thing on split pixels — and that looks more typically like what's happening here (aside from the extra thinness).

**Comment by connorlr** [comment]
I think there is something else at play here then, because these are definitely not the same.

Atom 1.40.1
<img width="739" alt="Screen Shot 2019-10-24 at 2 03 10 PM" src="https://user-images.githubusercontent.com/50108264/67512691-5f278280-f667-11e9-917b-9b0e842a8ed5.png">

Atom 1.41.0
<img width="765" alt="Screen Shot 2019-10-24 at 2 05 22 PM" src="https://user-images.githubusercontent.com/50108264/67512695-6058af80-f667-11e9-8aef-b9986a36a3ea.png">

**Comment by tcbbd** [comment]
@brendanfalkowski  This workaround should work better than just setting font-weight to bold.

```css
atom-text-editor {
  -webkit-text-stroke-width: 0.35px;
}
```

**Comment by marnixkoops** [comment]
On Mac OS Catalina the following worked for me.

**Change the OS defaults:**

- Execute at the Terminal: `defaults write -g CGFontRenderingFontSmoothingDisabled -bool NO`
- Completely exit Atom
- Start Atom again

**Comment by miniak** [comment]
Electron 4 is not supported anymore, this bug is not present on Electron 5+
https://electronjs.org/blog/electron-7-0#end-of-support-for-4xy

**Comment by miniak** [comment]
The problem is that https://skia-review.googlesource.com/c/skia/+/157566 wasn't back-ported properly. This is missing:
```
diff --git a/third_party/skia/src/ports/SkFontHost_mac.cpp b/third_party/skia/src/ports/SkFontHost_mac.cpp
index 8be3f306c1b8b1bf6609d14798860ca763f824b1..48dbf2f51d03d03863f948cdfc1aaa2dee008b29 100644
--- a/third_party/skia/src/ports/SkFontHost_mac.cpp
+++ b/third_party/skia/src/ports/SkFontHost_mac.cpp
@@ -440,6 +440,9 @@ static SmoothBehavior smooth_behavior() {
                 if (r != g || r != b) {
                     return SmoothBehavior::subpixel;
                 }
+                if (noSmoothBitmap[x][y] != smoothPixel) {
+                    smoothBehavior = SmoothBehavior::some;
+                }
             }
         }
         return smoothBehavior;
```

**Comment by justinko** [comment]
@tcbbd Doesn't apply for the tree or tabs text ... but it's a nice temp fix. Thanks!

**Comment by stakable** [comment]
@justinko `atom-workspace atom-workspace {
  -webkit-text-stroke-width: .15px;
}` instead of `atom-text-editor`
